### PR TITLE
[WiP] Container schema migration API proposal

### PR DESCRIPTION
**Pull request by alpaix**
### Motivation
Application data stored in a Fluid container might become outdated as the data schema evolves with the application functionality. The goal of this PR is to propose a public 3P API allowing application developers to update container data objects before the container is fully loaded.

### Assumptions
1. Evolution vs versioning. This proposal doesn't cover the container versioning story. We assume an application developer can identify what changes are required via inspecting container objects. This common adaptive strategy is called "evolution".

### Requirements
1. The new data manipulation API should be consistent with the `fluid-static` paradigm.
2. The API should be strictly scoped only to changing data objects in the container.
3. The API should not expose internal entities, such as the root data object.

### Design
An application is provided with an optional `migrations` schema parameter to specify a data manipulation callback(s). The callback signature includes a snapshot of container objects and a factory method to create a new object.

```ts
  const schema = {
      initialObjects: {
          map: SharedMap,
          directory: SharedDirectory,
      },
      migrations: async (snapshot: LoadableObjectRecord, createObject: ObjectFactory) => {
          if ("directory" in snapshot === false) {
              const directory = await createObject(SharedDirectory);
              return { ...snapshot, directory};
          }
          return undefined;
      },
  };

  const { container } = await client.getContainer(containerId, schema);
```

### Tests
Tests are provided to showcase usage patterns of the data migration API, such as adding, removing, updating container objects, in `schemaMigrations.spec.ts`.

### CODE REVIEW

**Comment by alpaix** [comment]
@ChumpChief @skylerjokiel It'd be great to get your feedback on this proposal

**Comment by ChumpChief** [comment]
I like that this approach puts the migration logic in the hands of the customer, giving them liberty to write exactly the migration steps required.  There are a few design points that I think need to be ironed out though:
1. So far I don't think this approach offers any guarantee that the migrator is unique, or avoidance for races.  So if two clients load a document simultaneously, both see that `"directory"` needs to be created, and both do so, then we'll hit a race condition that results in two separate directories being created and operated upon.  We'll need some plan for race safety.
2. Similar to the first point, it seems possible that there might be clients connected to the collaboration session, using the old schema when the a new client joins.  What happens in that case?  E.g. is there some approach to kick out the old clients, what if they have ops in flight on the old data that should also eventually get migrated, etc.  Between the first point and this one, I think it might be useful to see this strategy used in a sample (the unit tests are good, but I think seeing the whole flow of a migration will be important in evaluating it).
3. Migrations involving data "move", "merge", "split", etc. currently rely heavily on the extractability of the data within the data store.  This works fine for data structures like `SharedMap` which are pretty comprehensively accessible, but is not universally true (e.g. `ConsensusRegisterCollection` maintains (but hides from public view) sequence numbers for stored values, which are crucial to calculating eventually consistent values).  Successfully migrating some data stores will require an approach that can access their private information or otherwise mitigate the requirement.
4. This approach takes advantage of the `fluid-static` schema approach which has global knowledge of the schema from a centralized place -- that is, one authority can know all the needed migrations.  If we start to look at more nested scenarios (e.g. I've written part of my app as a DO and want to upgrade that DO's schema) I think we might need a way to delegate out migration steps to whoever is encapsulating schema knowledge.
5. Making this specific to the `fluid-static` schema is probably OK as a scoping measure, but ultimately we'll want a generic approach (either by pushing the `fluid-static` approach down into core runtime or by building a more generic implementation that `fluid-static` can sit on top of).  I'm curious what that more-generic approach might end up looking like in broad strokes, so we can compare this more-specific approach against it.

File: examples/hosts/app-integration/schema-upgrade/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# @fluid-example/app-integration-external-data
```
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/README.md:1]: Here and line 21, need to update to the correct package name (I think repo-policy-check is supposed to fail if the readme title doesn't match the package name?)

File: examples/hosts/app-integration/schema-upgrade/package.json [hunk 2]
```diff
@@ -0,0 +1,110 @@
+{
+  "name": "@fluid-example/app-integration-schema-upgrade",
+  "version": "0.57.0",
```
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/package.json:3]: Should be 0.58.1000, also dependencies need updated to their respective latest

File: examples/hosts/app-integration/schema-upgrade/package.json [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  "description": "Using external data to initialize the container state and serialize out afterwards.",
+  "homepage": "https://fluidframework.com",
+  "repository": "https://github.com/microsoft/FluidFramework",
```
Comment (ChumpChief) [inline examples/hosts/app-integration/schema-upgrade/package.json:6]: This probably needs to be updated to align with #9410

**Review by ChumpChief — COMMENTED** [review]
Added a couple comments - I'd suggest splitting the demo work to its own separate PR, since I think with these comments addressed we can go ahead and merge it whereas the fluid-static changes probably need more discussion.

**Comment by alpaix** [comment]
Abandoning in favor of a different approach based on export/import capabilities.

**Pull request closed without merging**